### PR TITLE
[WIP] Add Apache Parquet support to RAPIDS CUDF

### PR DESCRIPTION
**Pull request by felipeblazing**
This PR adds the parquet reader to cudf. Will be filling in more details when we are sure its ready for review here.

We will need help making the Python bindings for this new api. We know that process has changed with the move from cffi to cython and aren't aware of how to do this at the moment.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by kkraus14** [comment]
@felipeblazing given the scope of this PR I'm going to suggest we do the libcudf and python cudf changes in a followup PR.

**Comment by felipeblazing** [comment]
rerun tests

**Comment by felipeblazing** [comment]
rerun tests

**Comment by jrhemstad** [comment]
This is pretty monolithic. @felipeblazing is there any way this can be broken into more consumable parts? 

As it is, I don't even know where to begin reviewing this.

**Comment by aucahuasi** [comment]
@jrhemstad Thanks for the feedback.
The entry point would be cudf/io_functions.hpp then all the implementation is in src/io/parquet.
What would you need in order to start the review process?

**Comment by jrhemstad** [comment]
> @jrhemstad Thanks for the feedback.
> The entry point would be cudf/io_functions.hpp then all the implementation is in src/io/parquet.
> What would you need in order to start the review process?

Perhaps a live review in a WebEx would be a good place to start.

14,000 lines is just a lot to look through...

**Comment by felipeblazing** [comment]
> > @jrhemstad Thanks for the feedback.
> > The entry point would be cudf/io_functions.hpp then all the implementation is in src/io/parquet.
> > What would you need in order to start the review process?
> 
> Perhaps a live review in a WebEx would be a good place to start.
> 
> 14,000 lines is just a lot to look through...

We would be happy to do that. To be clear we kept  the same functions, files, and flow that are found in Parquet. We merely made changes to it so that some of the more costly operations occur on the GPU.

File: cpp/include/cudf/io_functions.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
+
```
Comment (vuule) [inline cpp/include/cudf/io_functions.hpp]: Why not merge this file with io_functions.h?

File: cpp/tests/io/parquet/api-test.cu [hunk 2]
```diff
@@ -0,0 +1,463 @@
+/*
```
Comment (vuule) [inline cpp/tests/io/parquet/api-test.cu]: I don't see any CUDA code in this file. Should it be a .cpp file instead?

File: cpp/tests/io/parquet/utils.cu [hunk 3]
```diff
@@ -0,0 +1,83 @@
+
```
Comment (vuule) [inline cpp/tests/io/parquet/utils.cu]: This file has .cu extension but does not contain any CUDA kernels. Should be .cpp AFAIK.

File: cpp/tests/io/parquet/utils.cuh [hunk 4]
```diff
@@ -0,0 +1,392 @@
+
```
Comment (vuule) [inline cpp/tests/io/parquet/utils.cuh]: Similarly, this should be a cpp header.

File: cpp/src/io/parquet/column_reader.cu [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+    template <>                                                                \
+    struct ParquetTraits<ParquetType> {                                        \
+        static constexpr gdf_dtype gdfDType = GdfDType;                        \
+    }
+
+TYPE_TRAITS_FACTORY(::parquet::BooleanType, GDF_INT8);
+TYPE_TRAITS_FACTORY(::parquet::Int32Type, GDF_INT32);
+TYPE_TRAITS_FACTORY(::parquet::Int64Type, GDF_INT64);
+TYPE_TRAITS_FACTORY(::parquet::FloatType, GDF_FLOAT32);
+TYPE_TRAITS_FACTORY(::parquet::DoubleType, GDF_FLOAT64);
```
Comment (vuule) [inline cpp/src/io/parquet/column_reader.cu]: Does this PR include BYTE_ARRAY support? I don't see that type anywhere in the diff.

File: cpp/src/io/parquet/api.cpp [hunk 6]
```diff
@@ -0,0 +1,779 @@
...
+}
+
+gdf_error
+_ProcessParquetReaderJobs(std::vector<ParquetReaderJob> &jobs) {
+
+    std::mutex lock;
+    int        job_index     = 0;
+    gdf_error  gdf_error_out = GDF_SUCCESS;
+
+    int num_threads = std::thread::hardware_concurrency();
```
Comment (vuule) [inline cpp/src/io/parquet/api.cpp]: With CUDA, this CPU side parallelization should become irrelevant. We should be able to get the same perf with a single thread launching CUDA kernels. Not sure if this is the case ATM.

**Comment by harrism** [comment]
@mjsamoht is this PR to be replaced with an alternative? If so, can we close this one?

**Comment by mjsamoht** [comment]
Yes, we're working on an alternative implementation that is mostly running on GPU.  We can close this.

**Pull request closed without merging**